hazelcast-jet-core/src/main/java/com/hazelcast/jet/Traversers.java

```diff
@@ -218,4 +229,30 @@ public void accept(T t) {
             nextItem = t;
         }
     }
+
+    private static class SingletonTraverser<T> implements Traverser<T> {
+        private Object item;
+
+        SingletonTraverser(@Nonnull T item) {
+            this.item = item;
+        }
+
+        @Override
+        public T next() {
+            try {
+                return (T) item;
+            } finally {
+                item = null;
+            }
+        }
+
+        @Nonnull @Override
+        // an optimized version to map in-place
```

Comment:
Yeah, you can't reuse the old traverser after mapping :). We have the same problem in `AppendableTraverser.append()` - it won't fail, but it will change behavior.

Maybe we can document that the methods on the Traverser can destroy the original traverser. We never use the traverser on which we called any method.

Or we can just drop this optimization, the gain is minor. j.u.s.Stream also doesn't support this. They throw `IllegalStateException` if you try.

Reply:
I'd be fine with explicitly documenting that calling any transformation method on the traverser invalidates it.